File: src/extras-attica/libtopic.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_Attica__Topic
#include <QDateTime>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <topic.h>
#include "libtopic.h"
#include "libtopic.hxx"

Attica__Topic* Attica__Topic_new() {
    return new Attica::Topic();
}

Attica__Topic* Attica__Topic_new2(const Attica__Topic* other) {
    return new Attica::Topic(*other);
}

void Attica__Topic_OperatorAssign(Attica__Topic* self, const Attica__Topic* other) {
    self->operator=(*other);
}

void Attica__Topic_SetId(Attica__Topic* self, const libqt_string id) {
    QString id_QString = QString::fromUtf8(id.data, id.len);
    self->setId(id_QString);
}

libqt_string Attica__Topic_Id(const Attica__Topic* self) {
    QString _ret = self->id();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Topic_SetForumId(Attica__Topic* self, const libqt_string forumId) {
    QString forumId_QString = QString::fromUtf8(forumId.data, forumId.len);
    self->setForumId(forumId_QString);
}

libqt_string Attica__Topic_ForumId(const Attica__Topic* self) {
    QString _ret = self->forumId();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Topic_SetUser(Attica__Topic* self, const libqt_string user) {
    QString user_QString = QString::fromUtf8(user.data, user.len);
    self->setUser(user_QString);
}

libqt_string Attica__Topic_User(const Attica__Topic* self) {
    QString _ret = self->user();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Topic_SetDate(Attica__Topic* self, const QDateTime* date) {
    self->setDate(*date);
}

QDateTime* Attica__Topic_Date(const Attica__Topic* self) {
    return new QDateTime(self->date());
}

void Attica__Topic_SetSubject(Attica__Topic* self, const libqt_string subject) {
    QString subject_QString = QString::fromUtf8(subject.data, subject.len);
    self->setSubject(subject_QString);
}

libqt_string Attica__Topic_Subject(const Attica__Topic* self) {
    QString _ret = self->subject();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Topic_SetContent(Attica__Topic* self, const libqt_string content) {
    QString content_QString = QString::fromUtf8(content.data, content.len);
    self->setContent(content_QString);
}

libqt_string Attica__Topic_Content(const Attica__Topic* self) {
    QString _ret = self->content();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Topic_SetComments(Attica__Topic* self, const int comments) {
    self->setComments(static_cast<const int>(comments));
}

int Attica__Topic_Comments(const Attica__Topic* self) {
    return self->comments();
}

bool Attica__Topic_IsValid(const Attica__Topic* self) {
    return self->isValid();
}

void Attica__Topic_Delete(Attica__Topic* self) {
    delete self;
}

File: src_original/extras-attica/libtopic.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/attica-topic.html
pub const attica__topic = struct {
    /// New constructs a new Attica::Topic object.
    ///
    ///
    pub fn New() QtC.Attica__Topic {
        return qtc.Attica__Topic_new();
    }

    /// New2 constructs a new Attica::Topic object.
    ///
    /// ``` other: QtC.Attica__Topic ```
    pub fn New2(other: ?*anyopaque) QtC.Attica__Topic {
        return qtc.Attica__Topic_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#operator-eq)
    ///
    /// ``` self: QtC.Attica__Topic, other: QtC.Attica__Topic ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.Attica__Topic_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setId)
    ///
    /// ``` self: QtC.Attica__Topic, id: []const u8 ```
    pub fn SetId(self: ?*anyopaque, id: []const u8) void {
        const id_str = qtc.libqt_string{
            .len = id.len,
            .data = id.ptr,
        };
        qtc.Attica__Topic_SetId(@ptrCast(self), id_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#id)
    ///
    /// ``` self: QtC.Attica__Topic, allocator: std.mem.Allocator ```
    pub fn Id(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Topic_Id(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::topic.Id: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setForumId)
    ///
    /// ``` self: QtC.Attica__Topic, forumId: []const u8 ```
    pub fn SetForumId(self: ?*anyopaque, forumId: []const u8) void {
        const forumId_str = qtc.libqt_string{
            .len = forumId.len,
            .data = forumId.ptr,
        };
        qtc.Attica__Topic_SetForumId(@ptrCast(self), forumId_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#forumId)
    ///
    /// ``` self: QtC.Attica__Topic, allocator: std.mem.Allocator ```
    pub fn ForumId(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Topic_ForumId(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::topic.ForumId: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setUser)
    ///
    /// ``` self: QtC.Attica__Topic, user: []const u8 ```
    pub fn SetUser(self: ?*anyopaque, user: []const u8) void {
        const user_str = qtc.libqt_string{
            .len = user.len,
            .data = user.ptr,
        };
        qtc.Attica__Topic_SetUser(@ptrCast(self), user_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#user)
    ///
    /// ``` self: QtC.Attica__Topic, allocator: std.mem.Allocator ```
    pub fn User(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Topic_User(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::topic.User: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setDate)
    ///
    /// ``` self: QtC.Attica__Topic, date: QtC.QDateTime ```
    pub fn SetDate(self: ?*anyopaque, date: ?*anyopaque) void {
        qtc.Attica__Topic_SetDate(@ptrCast(self), @ptrCast(date));
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#date)
    ///
    /// ``` self: QtC.Attica__Topic ```
    pub fn Date(self: ?*anyopaque) QtC.QDateTime {
        return qtc.Attica__Topic_Date(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setSubject)
    ///
    /// ``` self: QtC.Attica__Topic, subject: []const u8 ```
    pub fn SetSubject(self: ?*anyopaque, subject: []const u8) void {
        const subject_str = qtc.libqt_string{
            .len = subject.len,
            .data = subject.ptr,
        };
        qtc.Attica__Topic_SetSubject(@ptrCast(self), subject_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#subject)
    ///
    /// ``` self: QtC.Attica__Topic, allocator: std.mem.Allocator ```
    pub fn Subject(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Topic_Subject(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::topic.Subject: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setContent)
    ///
    /// ``` self: QtC.Attica__Topic, content: []const u8 ```
    pub fn SetContent(self: ?*anyopaque, content: []const u8) void {
        const content_str = qtc.libqt_string{
            .len = content.len,
            .data = content.ptr,
        };
        qtc.Attica__Topic_SetContent(@ptrCast(self), content_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#content)
    ///
    /// ``` self: QtC.Attica__Topic, allocator: std.mem.Allocator ```
    pub fn Content(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Topic_Content(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::topic.Content: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#setComments)
    ///
    /// ``` self: QtC.Attica__Topic, comments: i32 ```
    pub fn SetComments(self: ?*anyopaque, comments: i32) void {
        qtc.Attica__Topic_SetComments(@ptrCast(self), @intCast(comments));
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#comments)
    ///
    /// ``` self: QtC.Attica__Topic ```
    pub fn Comments(self: ?*anyopaque) i32 {
        return qtc.Attica__Topic_Comments(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-topic.html#isValid)
    ///
    /// ``` self: QtC.Attica__Topic ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.Attica__Topic_IsValid(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.Attica__Topic ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.Attica__Topic_Delete(@ptrCast(self));
    }
};
